### ISSUE TITLE
AF_UNIX stopped working sometime during 1.19

### ISSUE THREAD

**Issue by mikrosk** [freemint/freemint#330]
As reported here: https://www.atari-forum.com/viewtopic.php?p=456097#p456097

Even the client/server test located in https://github.com/freemint/freemint/tree/master/tools/net-tools/tests doesn't work anymore.

**Comment by MedourMehdi** [comment]
I'm not an expert and I may be wrong but could it be related to mintlib?
Those lines seem wrong to me https://github.com/freemint/mintlib/blob/master/socket/bind.c#L48-L50

**Comment by th-otto** [comment]
_libc_unix_names is normally set when you run the program on mint, so that code won't be reached. Also, if the same binary runs onf kernel 1.15 but not on 1.19, it is unlikely to be a mintlib issue.

The difference however is that kernel 1.19 implements the Fbind() gemdos function, while 1.15 does not.

**Comment by mikrosk** [comment]
Since the bug report says working on 1.18, it must be even more recent than 1.15.

**Comment by th-otto** [comment]
I think the problem is due to overly complicated hashing of different indices. In https://github.com/freemint/freemint/blob/0ebb10dfde5a9e8e4e6cf62946d0e06b76c879d7/sys/ipc_unix.c#L128 an entry is made to the allundatas array when you create an AF_UNIX. When you bind() to a unix socket (ie, start the server). The index of that entry is actually written to a regular file (https://github.com/freemint/freemint/blob/0ebb10dfde5a9e8e4e6cf62946d0e06b76c879d7/sys/ipc_unix.c#L224)

However, when the client tres to connect to it, a completely different index is used to look up the server: https://github.com/freemint/freemint/blob/0ebb10dfde5a9e8e4e6cf62946d0e06b76c879d7/sys/ipc_unix_stream.c#L39-L44. The index that un_namei returns is just a checksum of the filename, not the index that un_lookup expects.

I haven't found yet where that regressed. But imho the caching in ipc_unix_cache is totally useless if you have to open the file anyway to read back the index that un_lookup needs.

**Comment by MedourMehdi** [comment]
Thanks, so if I understand well a possible resolution would be to use
sys_f_read (fd, sizeof (index), (char *) &index);
in place of 
un_cache_lookup (un.sun_path, index)
in the un_namei function?

**Comment by th-otto** [comment]
I think the regression is from https://github.com/freemint/freemint/commit/c675c4ce84865fdd0c74f84e9098b1e4042b5b28

There, the code to read the index back was replaced by the checksum code, but this is obviously wrong. I can't find a tag for the 1.18 release, but i guess it was before that commit.
I already fixed it locally, but would like to know what the original commit tried to fix. /dev/log is certainly not a unix socket?

**Comment by MedourMehdi** [comment]
I don't know. Whatever before you fixed it AF_UNIX never worked on 1.19. Plus it seems you resolved my mysql issue (and I think many other non working ports that I have to try again)
<img width="1032" alt="Capture d’écran 2024-07-26 à 15 44 37" src="https://github.com/user-attachments/assets/9da5bd68-0e44-453a-b585-4cbccc42443c">

**Comment by MedourMehdi** [comment]
The commit mentioned seems related to this package https://freemint.github.io/sparemint/sparemint/html/packages/initscripts.html

**Comment by mikrosk** [comment]
> I can't find a tag for the 1.18 release, but i guess it was before that commit.

Indeed, c675c4c is from 9.12.2013 while FreeMiNT 1.18 was released on 17.3.2013 (and its branch was created a month before that).

**Comment by MedourMehdi** [comment]
I tested the example from AF - It works now.
[mint_af_unix_srv_client.zip](https://github.com/user-attachments/files/16535752/mint_af_unix_srv_client.zip)
<img width="539" alt="Capture d’écran 2024-08-08 à 01 52 27" src="https://github.com/user-attachments/assets/65ae164a-e1ed-444c-84e3-4591ee35c899">

**Comment by mikrosk** [comment]
Before closing this, it would be great to reproduce the original bug report which c675c4c tried to fix. I tracked it down to https://mikro.naprvyraz.sk/mint/201311/msg00002.html (first two messages) and https://mikro.naprvyraz.sk/mint/201312/msg00005.html (rest of the discussion).

If I'm not mistaken, "starting the system loggers" mentioned in the first message comes from EasyMiNT.

Deeper analysis here: https://mikro.naprvyraz.sk/mint/201312/msg00106.html and here: https://mikro.naprvyraz.sk/mint/201312/msg00109.html (i.e. initlog is innocent, no need to recompile it).

**Comment by mikrosk** [comment]
Maybe @ps68060 can remember more details?

**Comment by DavidGZ** [comment]
Actually I've seen puzzled this last two days how booting stops after the "Starting system loggers:" message, but it doesn't totally stall, booting resumes again around 10 seconds later, but this is on Aranym, on real hardware for sure it takes more time and it must look like it hanged completely.

**Comment by ps68060** [comment]
@mikrosk 
I don't recall this issue.  Did you mean me ?

**Comment by mikrosk** [comment]
Well, it was you who reported it: https://mikro.naprvyraz.sk/mint/201311/msg00002.html ... 10 years ago. ;)

**Comment by ps68060** [comment]
It may have slipped my memory from 2013 !!

I assumed that had been fixed shortly afterwards.

**Comment by MedourMehdi** [comment]
Is init.d still installed by easymint? If so who maintains it?

**Comment by mikrosk** [comment]
`/etc/init.d` is a directory, I guess you mean `syslogd` ?

**Comment by th-otto** [comment]
>  but this is on Aranym, on real hardware for sure it takes more time and it must look like it hanged completely.

I don't think this can't be tested reliable on Aranym (atleast not on a hostfs), since mknod (required to create the socket file) is not implemented there (https://github.com/aranym/aranym/blob/39830077d7add20802d0318fba5a36408d729b6e/src/natfeat/hostfs.cpp#L534-L537)

**Comment by MedourMehdi** [comment]
> Well, it was you who reported it: https://mikro.naprvyraz.sk/mint/201311/msg00002.html ... 10 years ago. ;)

I mean sysv, sorry, , due to professional deformation...

**Comment by DavidGZ** [comment]
> 
> I don't think this can't be tested reliable on Aranym (atleast not on a hostfs), since mknod (required to create the socket file) is not implemented there (https://github.com/aranym/aranym/blob/39830077d7add20802d0318fba5a36408d729b6e/src/natfeat/hostfs.cpp#L534-L537)

In my case I have EasyMiNT distribution installed in a ext2 partition inside an image file.

**Comment by MedourMehdi** [comment]
> `/etc/init.d` is a directory, I guess you mean `syslogd` ?

Do someone know from witch package syslogd comes?
Can't find it in sparemint base packages
https://freemint.github.io/sparemint/sparemint/html/packages/initscripts.html

**Comment by mikrosk** [comment]
here: https://freemint.github.io/sparemint/sparemint/html/packages/sysklogd.html

**Comment by MedourMehdi** [comment]
I tested syslogd from sparemint, it start in background with this error and don't seem to log it in /var/log/messages
"network logging disabled"
I've quickly built latest version of [sysklogd](https://github.com/troglobit/sysklogd).
It works as expected on Aranym + hostfs (c.f. screenshot)
Btw as a reminder we lack of 
SOCK_CLOEXEC 
SOCK_NONBLOCK
O_CLOEXEC
but it's an other story

I guess we need an easymint install with the latest kernel.

Sparemint
<img width="674" alt="Capture d’écran 2024-08-08 à 22 40 01" src="https://github.com/user-attachments/assets/957a4b94-1dd3-4632-a23e-63c7c5b0359b">

Latest sysklogd
<img width="654" alt="Capture d’écran 2024-08-08 à 22 39 27" src="https://github.com/user-attachments/assets/e059bbad-3f93-4aa6-8dc1-32183786cb34">

<img width="1022" alt="Capture d’écran 2024-08-08 à 22 22 06" src="https://github.com/user-attachments/assets/5126c66b-aa74-400c-b870-8eceb7c350d9">

**Comment by MedourMehdi** [comment]
Syslogd from sparemint create a socket located at /pipe/log by default.
If you pass the arg -p /tmp/sys.sock it symlink this file to /pipe/log.

The latest version don't symlink anything, it just create the socket via option -p as requested plus listen the *:514 udp port

**Comment by th-otto** [comment]
The error message "syslog/udp service unknown" probably means a missing /etc/services file.

**Comment by MedourMehdi** [comment]
> The error message "syslog/udp service unknown" probably means a missing /etc/services file.

True, thanks!

So with v1.3 there's no error and syslogd seems to log in /var/messages - attached the debug output for both 1.3 & 2.6

[syslogd_mint_debug.zip](https://github.com/user-attachments/files/16561489/syslogd_mint_debug.zip)

**Comment by ps68060** [comment]
I just updated the kernel from build 975... to d4e...

The boot stops at the system logger and I had to ctrl-c to continue through it and the dhcp and others.

Strange that this issue has returned after so long. 



⁣Peter​

**Comment by MedourMehdi** [comment]
Hi Peter,

Could you start the system logger manualy, ie with syslogd -d ?

**Comment by ps68060** [comment]
Isn't it the same version of syslogd that is working with the previous kernel ?

Not sure which version it is but it's from 2010. MintLib 0.59.1

⁣Peter​

**Comment by ps68060** [comment]
It's version 1.5.0

⁣Peter​

**Comment by MedourMehdi** [comment]
No, You should have a binary in /usr/sbin/syslogd who listen udp port 514 & socket file /pipe/log.

As you said it hangs at this level when you boot & you have to do ctrl-c in order to continue it could be useful to know if it start via tos2win bash command, i.e. if you execute /usr/sbin/syslogd -d (-d for debug) in order to check what error you get (or not)

**Comment by ps68060** [comment]
It won't boot with the latest kernel so I'm not able to do any syslogd tests.

⁣Peter​

**Comment by MedourMehdi** [comment]
> It won't boot with the latest kernel so I'm not able to do any syslogd tests. ⁣Peter​
> […](#)
> On 10 Aug 2024, 18:58, at 18:58, Medour Mehdi ***@***.***> wrote: No, You should have a binary in /usr/sbin/syslogd who listen udp port 514 & socket file /pipe/log. As you said it hangs at this level when you boot & you have to do ctrl-c in order to continue it could be useful to know if it start via tos2win bash command, i.e. if you execute /usr/sbin/syslogd -d (-d for debug) in order to check what error you get (or not) -- Reply to this email directly or view it on GitHub: [#330 (comment)](https://github.com/freemint/freemint/issues/330#issuecomment-2282227858) You are receiving this because you were mentioned. Message ID: ***@***.***>

A ok! also using ctrl-c?

**Comment by ps68060** [comment]
After several attempts with long waits before hitting ctrl-c on each, it booted. 

syslogd -d 

First show any errors.

⁣Peter​

**Comment by DavidGZ** [comment]
> 
> Could you start the system logger manualy, ie with syslogd -d ?

@mikrosk linked above some mailing list emails from 2013, reading the thread I don't know whether it's clear that syslogd stops the boot process.

I get this with syslogd 1.3:

root@aranym040:~> syslogd -d
listening on syslog UDP port.
Allocated parts table for 1024 file descriptors.
Starting.
Called init.
Called allocate_log, nlogs = -1.
cfline(*.info;mail.none;authpriv.none                           /var/log/messages)
symbolic name: info ==> 6
symbolic name: none ==> 16
symbolic name: mail ==> 16
symbolic name: none ==> 16
symbolic name: authpriv ==> 80
leading char in action: /
filename: /var/log/messages
Called allocate_log, nlogs = 0.
cfline(authpriv.*                                               /var/log/secure)
symbolic name: * ==> 255
symbolic name: authpriv ==> 80
leading char in action: /
filename: /var/log/secure
Called allocate_log, nlogs = 1.
cfline(mail.*                                                   /var/log/maillog)
symbolic name: * ==> 255
symbolic name: mail ==> 16
leading char in action: /
filename: /var/log/maillog
Called allocate_log, nlogs = 2.
cfline(*.emerg                                                  *)
symbolic name: emerg ==> 0
leading char in action: *
write-all
Called allocate_log, nlogs = 3.
cfline(uucp,news.crit                                           /var/log/spooler)
symbolic name: crit ==> 2
symbolic name: uucp ==> 64
symbolic name: news ==> 56
leading char in action: /
filename: /var/log/spooler
 0: 7F 7F  X 7F 7F 7F 7F 7F 7F 7F  X 7F 7F 7F 7F 7F 7F 7F 7F 7F 7F 7F 7F 7F  X FILE: /var/log/messages 
 1:  X  X  X  X  X  X  X  X  X  X FF  X  X  X  X  X  X  X  X  X  X  X  X  X  X FILE: /var/log/secure
 2:  X  X FF  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X FILE: /var/log/maillog
 3:  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  X WALL:
 4:  X  X  X  X  X  X  X  7  7  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X  X FILE: /var/log/spooler
logmsg: pri 56, flags 4, from aranym040, msg syslogd 1.3-3: restart.
Called fprintlog, logging to FILE /var/log/messages
syslogd: restarted.
Debugging disabled, SIGUSR1 to turn on debugging.

**Comment by MedourMehdi** [comment]
So could it be finaly initlog as initialy suspected?

**Comment by DavidGZ** [comment]
> So could it be finaly initlog as initialy suspected?

I don't know. I've attached two scripts (I've added to them the "txt" extensión to be able to upload them). 

syslogd is run from "syslog" script, it's done calling a function defined in the "functions" script, the "daemon()" function.
The boot process stalls inside that function, line 136 inside "functions.txt" file:

`	[ $? = 0 ] && success $"$base startup" || failure $"$base startup"`

I have very little experience reading scripts, if someone more trained could confirm what that line is doing.

[functions.txt](https://github.com/user-attachments/files/16573218/functions.txt)
[syslog.txt](https://github.com/user-attachments/files/16573219/syslog.txt)

**Comment by th-otto** [comment]
The line above just prints the exit status. The program itself is run using
```
initlog -q -c syslogd
```
What i can't find in the scripts is that `initlog`. Is that an executable, or maybe some function defined in `/etc/sysconfig/init`?

**Comment by MedourMehdi** [comment]
> The line above just prints the exit status. The program itself is run using
> 
> ```
> initlog -q -c syslogd
> ```
> 
> What i can't find in the scripts is that `initlog`. Is that an executable, or maybe some function defined in `/etc/sysconfig/init`?

It's started from etc/rc.d/init.d/functions who is called by etc/rc.d/rc

# rc.single     This file is executed by init when it goes into runlevel
#               1, which is the administrative state. It kills all
#               deamons and then puts the system into single user mode.
#               Note that the file systems are kept mounted.

**Comment by DavidGZ** [comment]
> The line above just prints the exit status. The program itself is run using
> 
> ```
> initlog -q -c syslogd
> ```

Running this command from a bash shell doesn't through any error.

> 
> What i can't find in the scripts is that `initlog`. Is that an executable, or maybe some function defined in `/etc/sysconfig/init`?

It's a binary, It's installed in /sbin. This the [rpm package](https://freemint.github.io/sparemint/sparemint/html/packages/initscripts.html) [1] that installs it, there seem to be not source code in the source package.


[1] https://freemint.github.io/sparemint/sparemint/html/packages/initscripts.html

**Comment by th-otto** [comment]
Hm, bad. Most of the files there are only scripts, but i also can't find a source of it. Not even BSD systems seem to use that initscripts package anymore.

**Comment by ps68060** [comment]
I was thinking I might chop search back through the releases to see when the issue reappeared.

I noticed the release files have recently  increased from 9 MB to 13MB.

https://tho-otto.de/snapshots/freemint/bootable/

⁣Peter​

**Comment by th-otto** [comment]
Yes, that increase is mostly from the ext2 binaries which have been added.

**Comment by ps68060** [comment]
Build D01 from 08/07/2024 works.
Build B22 from 29/07/2024 doesn't work. 

I will continue to narrow it down.

⁣Peter​

**Comment by MedourMehdi** [comment]
I guess if initlog & syslog work from bash it comes from rc script

**Comment by DavidGZ** [comment]
> I was thinking I might chop search back through the releases to see when the issue reappeared.

I don't think this is necessary, the issue reappeared with the fix for the issue this thread is about. The latest posts are about the regression after the commit with the fix [c4eb6be](https://github.com/freemint/freemint/commit/c4eb6beb8d8bfae5e8be3b422b77a2d03aca07f8)

**Comment by MedourMehdi** [comment]
> The line above just prints the exit status. The program itself is run using
> 
> ```
> initlog -q -c syslogd
> ```
> 
> What i can't find in the scripts is that `initlog`. Is that an executable, or maybe some function defined in `/etc/sysconfig/init`?

Are you sure?
In the old [threads](https://github.com/freemint/freemint/issues/330#issuecomment-2275069369) it was said initlog -s?

**Comment by DavidGZ** [comment]
> Hm, bad. Most of the files there are only scripts, but i also can't find a source of it. Not even BSD systems seem to use that initscripts package anymore.
> 
> Edit: found some older sparemint packages, buts its the same everywhere: there is only a binary in the source packages. And its binary identical in all cases (143842 bytes), compiled against an ancient mintlib (0.57.1)
> 
> Edit2: the oldest i could find is from here: https://ftp.redhat.com/pub/redhat/linux/enterprise/4/en/os/i386/SRPMS/ but i don't know if that is really the same (the version number atleast is completely different)
> 
> What irritates me, that this binary in turn seems to use `syslog()`, something that the daemon in question is just setting up...

I've found [this repository](https://github.com/fedora-sysv/initscripts/commit/c3695a8497dcbc4f5dc3348f9dd0f4ffb01badee)

As far as I can see initlog first appear in intscripts with version 3.84.

**Comment by MedourMehdi** [comment]
We don't know what is hanging.
The process is /sbin/init -> /etc/rc (symlink) -> Loop scripts in /etc/init.d (symlink)
I think we need to reproduce what is happening because we don't know what is the incriminated command line.

**Comment by DavidGZ** [comment]
In the emails from 2013 Alan said that what it is hanging is initlog with the argument -s, this can be reproduced with kernels after this commit [c4eb6be](https://github.com/freemint/freemint/commit/c4eb6beb8d8bfae5e8be3b422b77a2d03aca07f8).

The line that is stopping the boot process is the one pointed below, from the code posted above inside "functions" file:

```
# Log that something succeeded
success() {
  if [ -z "$IN_INITLOG" ]; then
     initlog $INITLOG_ARGS -n $0 -s "$1" -e 1   <-----------
  else
     # silly hack to avoid EPIPE killing rc.sysinit
     trap "" SIGPIPE
     echo "$INITLOG_ARGS -n $0 -s \"$1\" -e 1" >&21
     trap - SIGPIPE
  fi
  [ "$BOOTUP" != "verbose" ] && echo_success
  return 0
}
```
In one of my post above I didn't notice that "success()" was also function called from inside "daemon()"

The command `initlog -n test -s "test" -e 1` doesn't return, this can be tested from the command line.

What I think it's nor clear is whether the culprit is initlog or the kernel. Alan first thought that it was initlog but later changed his opinion a pointed to the kernel.

**Comment by th-otto** [comment]
Ah, missed that one, too. 

Maybe his patches to the kernel fixed the initlog hang, but they were definitely wrong, since it breaks all other applications using AF_UNIX sockets. So we definitely would need some source of initlog, so we can fix the problem there.

**Comment by mikrosk** [comment]
@th-otto are you sure that it is really initlog's issue? Alan explicitly mentioned in the email thread that it is innocent here and it is purely a kernel problem.

**Comment by DavidGZ** [comment]
I've been looking further into this issue. The problem is that SpareMiNT's syslogd version doesn't create any socket, instead it creates a pipe with a symlink in /dev (/dev/log). initlog doesn't know anything about it and it tries to connect() to /dev/log like if it was a socket. The kernel handles the connect() call for a socket and it ends up trying to read the index of the socket which is stored in the file bonded to it by the bind() function [[1]](https://github.com/freemint/freemint/blob/0c8bc27e427d5b2097b5db07d185ca5cdffa44e2/sys/ipc_unix_cache.c#L178), because nothing is read from the pipe the process stalls here [[2]](https://github.com/freemint/freemint/blob/0c8bc27e427d5b2097b5db07d185ca5cdffa44e2/sys/pipefs.c#L1019).

[1] https://github.com/freemint/freemint/blob/0c8bc27e427d5b2097b5db07d185ca5cdffa44e2/sys/ipc_unix_cache.c#L178
[2] https://github.com/freemint/freemint/blob/0c8bc27e427d5b2097b5db07d185ca5cdffa44e2/sys/pipefs.c#L1019

**Comment by mikrosk** [comment]
Hm, re-reading Alan's original fix: https://github.com/freemint/freemint/commit/c675c4ce84865fdd0c74f84e9098b1e4042b5b28 doesn't sound that much different, does it. Do you understand what was the idea behind the checksum then?

**Comment by MedourMehdi** [comment]
Whatever the examples and all AF_LOCAL sockets were broken by this temporary fix regarding one old binary.
@DavidGZ may be you could test the latest syslogd I've build to check if you get the same behavior?

**Comment by th-otto** [comment]
> it. Do you understand what was the idea behind the checksum then?

ipc_unix_cache has only a limited number of cache entries for socket files (https://github.com/freemint/freemint/blob/0c8bc27e427d5b2097b5db07d185ca5cdffa44e2/sys/ipc_unix_cache.c#L21) that are created this way. The checksum is used to identify which are currently used so they can be closed/opened as needed.

**Comment by th-otto** [comment]
> The problem is that SpareMiNT's syslogd version doesn't create any socket, instead it creates a pipe with a symlink in /dev (/dev/log). initlog doesn't know anything about it and it tries to connect() to /dev/log like if it was a socket.

But i wonder why bind() does not already block, when it tries to write the index to that pipe?

Anyway, that really  sounds like that we  have to replace syslogd/initlog from sparemint somehow. And warn users about not using the old versions with newer kernels.

Maybe we can also take some precautions, by checking whether that is really a regular file and not a pipe. And ideally, we could also change the /pipe pseudo filesystem to allow not only creation of fifos, but also creation of real socket files. All that mess was imho only done because you can't create socket files on non-ext2 partitions.

**Comment by mikrosk** [comment]
> Anyway, that really sounds like that we have to replace syslogd/initlog from sparemint somehow. And warn users about not using the old versions with newer kernels.

This will be quite hard because it's the core package of EasyMiNT. We'd need to provide at least a RPM to upgrade seamlessly. Maybe even attach that RPM to the archives.

**Comment by th-otto** [comment]
I just meant to provide the binaries in the snapshots, so they can be replaced. Providing a rpm would be difficult, since all RPMs nowadays produce a format that is not supported by the rpm binary of EasyMint. Not to mention that it is almost impossible to setup an environment that produces the same meta data.

**Comment by mikrosk** [comment]
Yeah, my idea was to use EasyMiNT's environment, including `rpm`. But sure, providing the executable alone is also an option.

**Comment by DavidGZ** [comment]
> Hm, re-reading Alan's original fix: [c675c4c](https://github.com/freemint/freemint/commit/c675c4ce84865fdd0c74f84e9098b1e4042b5b28) doesn't sound that much different, does it.

Yes, in the commit message is the problem quite well explained. I didn't totally understand it when I read it the first time, it would have saved me some time while investigating this. But it's strange that syslog and initlog were not blamed, although the fix avoids the process to stall they are still malfunctioning.

> Do you understand what was the idea behind the checksum then?

The index to identify an entry in the cache was and it's now an arbitrary number, so I guess the idea was to generate an unique index based on the path of the socket file, then you don't need anymore to store it inside the file and you can avoid the read() action when you were looking up in the cache.

I wonder why this wasn't working for other processes.

**Comment by DavidGZ** [comment]
> Whatever the examples and all AF_LOCAL sockets were broken by this temporary fix regarding one old binary. @DavidGZ may be could you test the latest syslogd I've build to check if you get the same behavior?

I tested it and it doesn't behave the same, now syslogd creates a socket in /var/run/log. The problem now it's that SpareMiNT's  initlog looks for the socket in /dev/log.

Also notice that older original syslogd were creating the socket in /dev/log, MiNT's version was patched to create the pipe instead. I think the reason could be that bind() can't create a file in /dev as @th-otto has already said, or that when the application was ported sockets were not integrated in the kernel yet.

**Comment by DavidGZ** [comment]
> It's curious that those two processes have never worked, not before neither after the fix.

I must correct myself syslogd/initlog seems to work correctly after Alan's fix, initlog tries to connect() to /dev/log as it was a socket but still deals OK with the pipe to pass messages.

> Anyway, that really sounds like that we have to replace syslogd/initlog from sparemint somehow. And warn users about not using the old versions with newer kernels.

I was thinking that maybe with just fixing syslogd to create a socket instead of a pipe would be enough. I was thinking about creating the socket in /tmp (/tmp/log) and a symlink in /dev pointing to it. This fixes the stall but initlog doesn't seem work. I think the problem is that /pipe/log is set in [MiNTlib](https://github.com/freemint/mintlib/blob/059189ca233d345fc85d628f8f6c62b2474a05bb/misc/syslog.c#L88) [1].

This means that there are more binaries out there expecting to deal with a pipe instead of a socket. So recompiling only syslogd and initlog wouldn't be enough.

Perhaps /pipe/log should stay and we should fix first initlog to don't do the connect() call to that pipe.


[1] https://github.com/freemint/mintlib/blob/059189ca233d345fc85d628f8f6c62b2474a05bb/misc/syslog.c#L88

**Comment by MedourMehdi** [comment]
I don't know if it's useful but there're options -p to tell the socket location and -a to add more sockets location.

**Comment by DavidGZ** [comment]
I've prepared a [commit](https://github.com/DavidGZ/freemint/commit/e7d963c05143e2f375661e7e83a142bab109b0b3) [1] which adds a check in the kernel to avoid what was producing the stalls during SpareMiNT's boot process. Please take a look, if nobody has any objection I'll push it during the next days.

Also I've patched initlog to don't try to do a connect() call to /dev/log. I've taken the sources from here [[2]](https://github.com/fedora-sysv/initscripts/tree/r6-40-2/src). I've used version 6.40.2 which was released around the same time latest MiNT's inilog was built. Remember there were no source code in the rpm package. I've attached in this post the binary, source code and new rpms for initscripts-1.13, where initlog is installed from.
[initlog.zip](https://github.com/user-attachments/files/16664287/initlog.zip)
[initlog-rpm.zip](https://github.com/user-attachments/files/16664288/initlog-rpm.zip)

There is still an issue, although everything loads correctly during the boot process after the "Starting system loggers" message the word FAILED is shown instead of OK. This has to do with the scripts, if syslogd returns anything else than 0 the FAILED message is shown. What I've seen is that syslogd forks and then it kills itself with a SIGINT signal, this is the return value that the script is reading. I'm confused about that other systems didn't have the same issue back then, probably I'm missing something but I was not able to see what it could be.

[1] https://github.com/DavidGZ/freemint/commit/e7d963c05143e2f375661e7e83a142bab109b0b3
[2] https://github.com/fedora-sysv/initscripts/tree/r6-40-2/src

**Comment by th-otto** [comment]
The commit looks ok to me. However, it will only prevent the hang, users still have to update to the new version of initlog to get it working.

BTW, how did you create the RPM? On real hardware using the sparemint versions?

**Comment by DavidGZ** [comment]
> The commit looks ok to me. However, it will only prevent the hang, users still have to update to the new version of initlog to get it working.
> 

At first I was thinking about keeping Alan's code and try to fix it, this would allow old initlog versions to work too. I can pick up again this idea.

> BTW, how did you create the RPM? On real hardware using the sparemint versions?

Yes, on Aranym with EasyMiNT .

**Comment by DavidGZ** [comment]
> There is still an issue, although everything loads correctly during the boot process after the "Starting system loggers" message the word FAILED is shown instead of OK. This has to do with the scripts, if syslogd returns anything else than 0 the FAILED message is shown. What I've seen is that syslogd forks and then it kills itself with a SIGINT signal, this is the return value that the script is reading. I'm confused about that other systems didn't have the same issue back then, probably I'm missing something but I was not able to see what it could be.

I've seen that the code with the SIGINT signal was removed short after and replaced with an exit(0) call.

**Comment by DavidGZ** [comment]
> > The commit looks ok to me. However, it will only prevent the hang, users still have to update to the new version of initlog to get it working.
> 
> At first I was thinking about keeping Alan's code and try to fix it, this would allow old initlog versions to work too. I can pick up again this idea.

At the end the old initlog works with the new patched kernel too, so the new initlog would only be necessary with the kernel versions that stall.

**Comment by DavidGZ** [comment]
> It's version 1.5.0 ⁣Peter​

@ps68060, do you remember from where did you get that version. Version in SpareMiNT site is 1.3-5.

**Comment by ps68060** [comment]
I'm not sure. It might have been Alan.

⁣Peter​

**Comment by DavidGZ** [comment]
I've just pushed [c148ce8](https://github.com/freemint/freemint/commit/c148ce827a2968538f360cdc31bb3e33f7fd5617).

If there is not objection I would remove the checksum calculation code in [[1]](https://github.com/freemint/freemint/blob/master/sys/ipc_unix_cache.c), right now it's not used for anything.


[1] https://github.com/freemint/freemint/blob/master/sys/ipc_unix_cache.c

**Comment by th-otto** [comment]
Sounds good. Maybe you can also pass along the xattr structure, so sys_f_xattr does not have to be called again.

**Comment by DavidGZ** [comment]
> Sounds good. Maybe you can also pass along the xattr structure, so sys_f_xattr does not have to be called again.

It's done and pushed, but when I tested it I forgot to copy the new compiled kernel with the changes, something is not working right, I'll fix it later.
